src/tactic/ring.lean

```diff
@@ -56,49 +69,86 @@ do α ← infer_type e,
    nc ← mk_instance_cache `(ℕ),
    using_new_ref ic $ λ r,
    using_new_ref nc $ λ nr,
-   prod.fst <$> state_t.run (reader_t.run m ⟨α, u, c, red, r, nr⟩) mk_buffer
+   using_new_ref mk_buffer $ λ atoms,
+   reader_t.run m ⟨α, u, c, red, r, nr, atoms⟩
 
+/-- Lift an instance cache tactic (probably from `norm_num`) to the `ring_m` monad. This version
+is abstract over the instance cache in question (either the ring `α`, or `ℕ` for exponents). -/
 @[inline] meta def ic_lift' (icf : cache → ref instance_cache) {α}
   (f : instance_cache → tactic (instance_cache × α)) : ring_m α :=
-⟨λ c, ⟨λ s, do
+⟨λ c, do
   let r := icf c,
   ic ← read_ref r,
   (ic', a) ← f ic,
-  write_ref r ic',
-  pure (a, s)⟩⟩
+  a <$ write_ref r ic'⟩
 
+/-- Lift an instance cache tactic (probably from `norm_num`) to the `ring_m` monad. This uses
+the instance cache corresponding to the ring `α`. -/
 @[inline] meta def ic_lift {α} : (instance_cache → tactic (instance_cache × α)) → ring_m α :=
 ic_lift' cache.ic
 
+/-- Lift an instance cache tactic (probably from `norm_num`) to the `ring_m` monad. This uses
+the instance cache corresponding to `ℕ`, which is used for computations in the exponent. -/
 @[inline] meta def nc_lift {α} : (instance_cache → tactic (instance_cache × α)) → ring_m α :=
 ic_lift' cache.nc
 
+/-- Apply a theorem that expects a `comm_semiring` instance. This is a special case of
+`ic_lift mk_app`, but it comes up often because `horner` and all its theorems have this assumption;
+it also does not require the tactic monad which improves access speed a bit. -/
 meta def cache.cs_app (c : cache) (n : name) : list expr → expr :=
 (@expr.const tt n [c.univ] c.α c.comm_semiring_inst).mk_app
 
+/-- Every expression in the language of commutative semirings can be viewed as a sum of monomials,
+where each monomial is a product of powers of atoms. We fix a global order on atoms (up to
+definitional equality), and then separate the terms according to their smallest atom. So the top
+level expression is `a * x^n + b` where `x` is the smallest atom and `n > 0` is a numeral, and
+`n` is maximal (so `a` contains at least one monomial not containing an `x`), and `b` contains no
+monomials with an `x` (hence all atoms in `b` are larger than `x`).
+
+If there is no `x` satisfying these constraints, then the expression must be a numeral. Even though
+we are working over rings, we allow rational constants when these can be interpreted in the ring,
+so we can solve problems like `x / 3 = 1 / 3 * x` even though these are not technically in the
+language of rings.
+
+These constraints ensure that there is a unique normal form for each ring expression, and so the
+algorithm is simply to calculate the normal form of each side and compare for equality.
+
+To allow us to efficiently pattern match on normal forms, we maintain this inductive type that
+holds a normalized expression together with its structure. We also -/
```

Comment:
"We also"?

Reply:
Not really sure what I intended to say there...